util/log: extend `logpb.Entry` with more fine grained fields

knz opened the pull request:
Needed for #65633 and #61378.

cc @cockroachdb/server

knz committed:
util/log: move the crdb-v1 test code to a separate file

Release note: None

knz committed:
util/log: use a datadriven test for the crdb-v1 format/parse test

This simplifies the test and makes it more extensible.

Release note: None

knz committed:
util/log: make `logpb.Entry` slightly more able to represent crdb-v2 entries

Prior to this change, we did not have a public data structure able to
represent the various fields available in a `crdb-v2` log entry: the
existing `logpb.Entry` was not able to distinguish structured and
non-structured entries, and did not have the ability to delimit the
message and the stack trace.

This patch extends `logpb.Entry` to make it more able to represent
`crdb-v2` entries, at least for the purpose of extending `debug
merge-log` towards conversion between logging formats.

Additionally, the patch adds a *best effort* attempt at populating
the new fields in the `crdb-v1` entry parser. This is best effort
because the crdb-v1 parser is lossy so we cannot faithfully
parse its entries reliably.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66096)
<!-- Reviewable:end -->

cameronnunez commented on pkg/util/log/format_crdb_v1_test.go:
```diff
@@ -0,0 +1,177 @@
+
+	// Verify the truncation logic for reading logs that are longer than the
+	// default scanner can handle.
+	preambleLength := len(formatEntry(severity.INFO, channel.DEV, t1, 0, "clog_test.go", 136, ``, ""))
```
Each instance of the string "clog_test.go" in `format_crdb_v1_test.go` and `testdata/crdb_v1` should be changed to the string "format_crdb_v1_test.go"

cameronnunez reviewed (changes requested):
After these minor changes, LGTM! Thanks for your work here.

knz reviewed (commented):
nit: if you are only requesting minor changes, then it's better to leave your review disposition as "neutral" instead of "changes requested".
With "changes requested" I have to wait until you complete the review before I can merge the PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66096#-:-MbQiHzu7TfrTAatJXAx:b-vbmgq6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez)

___
*[pkg/util/log/format_crdb_v1_test.go, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66096#-MbNLNYoUDQGWSMWUy4l-r3-32:-MbQhqDyE89Ucfj_1qfF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc0921781bafc52a555c48404759e4ad6e3654a/pkg/util/log/format_crdb_v1_test.go#L32)):*
<details><summary><i>Previously, cameronnunez (Cameron Nunez) wrote…</i></summary><blockquote>

```suggestion
func TestFormatCrdbV1Decode(t *testing.T) {
```

For consistency, I suggest a slight change to the function name here.
</blockquote></details>

Done.
___
*[pkg/util/log/format_crdb_v1_test.go, line 122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66096#-MbNLNYoUDQGWSMWUy4l-r3-122:-MbQhqyx-RO3JocBQWfC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc0921781bafc52a555c48404759e4ad6e3654a/pkg/util/log/format_crdb_v1_test.go#L122)):*
<details><summary><i>Previously, cameronnunez (Cameron Nunez) wrote…</i></summary><blockquote>

Each instance of the string "clog_test.go" in `format_crdb_v1_test.go` and `testdata/crdb_v1` should be changed to the string "format_crdb_v1_test.go"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz committed:
util/log: report the logging format at the start of new files

Release note (cli change): When log entries are written to disk,
the first few header lines written at the start of every new file
now report the configured logging format.

knz committed:
logconfig: fix the handling of crdb-v1 explicit config

Release note (bug fix): Previously, `--log='file-defaults: {format:
crdb-v1}'` was not handled properly. This has been fixed. This bug
existed since v21.1.0.

knz committed:
cli: report explicit log config in logs

This increases troubleshootability.

Release note: None

cameronnunez commented:
> nit: if you are only requesting minor changes, then it's better to leave your review disposition as "neutral" instead of "changes requested".
> With "changes requested" I have to wait until you complete the review before I can merge the PR.
Understood!

knz commented:
You still need to review and change your disposition.
![image](https://user-images.githubusercontent.com/642886/121057860-0404be00-c7c0-11eb-9754-a8d1cdb81153.png)

cameronnunez commented:
> You still need to review and change your disposition.
> ![image](https://user-images.githubusercontent.com/642886/121057860-0404be00-c7c0-11eb-9754-a8d1cdb81153.png)

👍 Done.

knz commented:
awesome, thank you!

bors r=cameronnunez
